src/app/editor/editor.component.ts

```diff
@@ -1,76 +1,77 @@
   selector: 'editor',
-  providers: [
-    RecordService,
-    JsonUtilService,
-    SchemaRefResolverService,
-    SchemaEnforcerService
-  ],
+  encapsulation: ViewEncapsulation.None,
   directives: [
-    ObjectArrayFieldComponent,
     ArrayInArrayFieldComponent,
-    StringArrayFieldComponent,
+    PrimitiveArrayFieldComponent,
+    PrimitiveFieldComponent,
+    ObjectArrayFieldComponent,
     ObjectFieldComponent
   ],
   pipes: [MapToIterablePipe, UnderscoreToSpacePipe],
+  providers: [
+    JsonUtilService,
+    RecordService,
+    RecordFixerService,
+  ],
   styles: [
     require('./editor.component.scss')
   ],
   template: require('./editor.component.html'),
-  encapsulation: ViewEncapsulation.None
 })
 export class EditorComponent extends AbstractTrackerComponent {
   // TODO: remove dummy
   jsonDoc: any = {};
   schema: any = {};
 
-  constructor(private recordService: RecordService, 
-    private jsonUtilService: JsonUtilService, 
-    private schemaRefResolverService: SchemaRefResolverService,
-    private schemaEnforcerService: SchemaEnforcerService) {
+  constructor(private recordService: RecordService,
+    private jsonUtilService: JsonUtilService,
+    private recordFixerService: RecordFixerService) {
     super();
+
     // FIXME: find a better way to make editor wait until we got the schema
     let record;
     
-    // FIXME: do not combine Promise and Observable
     this.recordService.fetchRecord('literature', '1404685')
       .flatMap(json => {
         record = json;
         return this.recordService.fetchSchema(json['$schema']);
-      }).toPromise().then(schema => {
-        // TODO: Remove this, when record comes from DB not from Elasticsearch and schema is more consistent
+      }).subscribe(schema => {
+        // TODO: Remove these delete, when record comes from DB not from Elasticsearch and schema is more consistent
         Object.keys(record).forEach((prop) => {
-          if (record[prop].constructor.name === 'Array' && schema[prop] == null) {
+          if (schema[prop] == null) {
             delete record[prop];
             console.log('not in schema => ', prop);
-          } else if (record[prop].constructor.name === 'Object' && schema[prop].properties == null) {
-            delete record[prop];
-            console.log('no .properties => ', prop);
           }
         });
-        // TODO: Remove this when the schema comes already resolved! (WARNING: a lot of http requests!)
-        return this.schemaRefResolverService.resolveRefs(schema);
-      }).then(resolvedSchema =>{
-        this.schema = resolvedSchema;
-        this.schemaEnforcerService.enforceSchema(record, resolvedSchema);
+        delete record['self']; // ingrone self
+
+        this.recordFixerService.fixRecord(record, schema);
         this.jsonDoc = this.jsonUtilService.flattenMARCJson(record);
-      });
+        this.schema = schema;
+      }, error => console.error(error));
+  }
+
+  onValueChange(event: any, key: string) {
```

Comment:
Kinda. So components that have this, they pass some value to children components and want to be notified when data is changed. So children emit the value change and propagates it to parent, then this function called.

In short this is a way to achieve two-way data binding.